lib/backend/watchersyncer/watchercache.go

```diff
@@ -0,0 +1,296 @@
+		// go routine (if we haven't already) and by sending deletes for the old resources that were
+		// not acknowledged by the List.  The oldResources will be empty after this call.
+		wc.dataIsNowSynced()
+
+		// And now start watching from the revision returned by the List.
+		w, err := wc.client.Watch(context.Background(), wc.resourceType.ListInterface, l.Revision)
+		if err != nil {
+			// Failed to create the watcher - we'll need to retry.  Sleep so that we don't
+			// tight loop.  Since we have just performed a list, we need a slightly longer
+			// delay to avoid overloading the datastore.  If the watcher keeps failing then
+			// we are effectively operating in a polling mode, so the interval should be a
+			// sensible polling interval.
+			wc.logger.WithError(err).Debug("Failed to create watcher")
+			time.Sleep(WatchPollInterval)
+			continue
+		}
+
+		// Store the watcher and exit back to the main event loop.
+		wc.watch = w
+		return
+	}
+}
+
+// dataIsNowSynced handles processing to finish synchronization.
+// If this watcher has never been synced then notify the main watcherSyncer that we've synced.
+// We may also need to send deleted messages for old resources that were not validated in the
+// resync (i.e. they must have since been deleted).
+func (wc *watcherCache) dataIsNowSynced() {
+	// If this is our first synced event then send a synced notification.  The main
+	// watcherSyncer code will send a Synced event when it has received synced events from
+	// each cache.
+	if !wc.hasSynced {
+		wc.logger.Info("Sending synced update")
+		wc.results <- api.InSync
+		wc.hasSynced = true
+	}
+
+	// If the watcher failed at any time, we end up recreating a watcher and storing off
+	// the current known resources for revalidation.  Now that we have finished the sync,
+	// any of the remaining resources that were not accounted for must have been deleted
+	// and we need to send deleted events for them.
+	numOldResources := len(wc.oldResources)
+	if numOldResources > 0 {
+		wc.logger.WithField("Num", numOldResources).Debug("Sending resync deletes")
+		updates := make([]api.Update, 0, len(wc.oldResources))
+		for _, r := range wc.oldResources {
+			updates = append(updates, api.Update{
+				UpdateType: api.UpdateTypeKVDeleted,
+				KVPair: model.KVPair{
+					Key: r.key,
+				},
+			})
+		}
+		wc.results <- updates
+	}
+	wc.oldResources = nil
+}
+
+// handleWatchEvent handles a watch event converting it required and fanning out
+// to the add/mod or delete processing as necessary.
+func (wc *watcherCache) handleWatchListEvent(kvp *model.KVPair) {
+	if wc.resourceType.UpdateProcessor == nil {
+		// No update processor - handle immediately.
+		wc.handleConvertedWatchEvent(kvp)
+		return
+	}
+
+	// We have an update processor so use that to convert the event data.
+	kvps, err := wc.resourceType.UpdateProcessor.Process(kvp)
+	for _, kvp := range kvps {
```

Comment:
it's a perfectly valid thing to have both errors and data returned in the same call.  It was my intended use case for partial conversions, yes.  I don't think I actually use that feature anywhere though - so perhaps remove it later if none of the concrete processors make use of it.

Reply:
I'll change if you really want me to, otherwise I'll leave it for now.  I think if it made the code path more complicated I'd omit it, but in some ways it actually makes it simpler.

Reply:
It's fine. I don't really like it but I'm not bothered enough to ask you to change.